server/aesdsocket.c

```diff
@@ -14,140 +14,288 @@
+    struct thread_data* thread_info = (struct thread_data *)thread_param;
+    inet_ntop(AF_INET,thread_info->client,clientIP,sizeof(clientIP));
+    bool recv_complete=false;
+    buf=malloc(BUFFER_SIZE*sizeof(char*));
+    while(!recv_complete){
+        buf=(char*)(realloc(buf,current_size+BUFFER_SIZE));
+        s_recv=recv(thread_info->socketfd,(buf+current_size),BUFFER_SIZE,0);
+        if(s_recv==-1){
+            syslog(LOG_USER, "Error while recieving");
+            perror("recv");
+            exit(-1);
+        }
+        size_recived=current_size+s_recv;
+        if(size_recived>0 && buf[size_recived-1]=='\n'){
+            recv_complete=true;
+            buf[size_recived-1]='\n';
+        }
+        current_size+=BUFFER_SIZE;
+    }
+    if(recv_complete){
+    ret=pthread_mutex_lock(&mutex); 
+    if(ret!=0){
+    syslog(LOG_USER,"Error while unlocking mutex");
+    exit(-1);
+} 
+    file_fd=open(FILE_PATH, O_RDWR | O_CREAT | O_APPEND ,0644);
+    if(file_fd == -1){
+        syslog(LOG_USER, "Unable to open file to read, Check the permission");
+        perror("open");
+        exit(-1);
+    }     
+    if(timer_alarm){
+        write_time_to_file(file_fd);
+        timer_alarm=false;    
+    }
+        nr=write(file_fd, buf, size_recived);
+        if(nr!=size_recived){
+            syslog(LOG_USER, "Writing to file not Successfull"); 
+            perror("write");
+            exit(-1);
+            }   
+           
+    }
+    char message[BUFFER_SIZE];
+    lseek(file_fd,0,SEEK_SET);
+    while((nr=read(file_fd,message,BUFFER_SIZE))!=0){
+        if(nr ==-1){
+            syslog(LOG_USER, "Reading from file not Successfull");   
+            exit(-1);
+        } 
+        s_send=send(thread_info->socketfd,message,(nr/sizeof(char)),0);
+        if(s_send<0){
+           syslog(LOG_USER, "Sending failed"); 
+        }        
+    }
+close(file_fd);  
+ret=pthread_mutex_unlock(&mutex);   
+if(ret!=0){
+    syslog(LOG_USER,"Error while unlocking mutex");
+    exit(-1);
+}
+thread_info->connection_complete_success =true;            
+}
+
 static void signal_handler (int signo)
 {
-    if(signo == SIGINT || signo == SIGTERM){  
+    int ret,fd;
+    struct thread_data* ele;
+    if(signo == SIGINT || signo == SIGTERM){   
     syslog (LOG_USER,"Caught Signal Exiting!\n");
-    remove(FILE_PATH);
-    exit (EXIT_SUCCESS);
+    close(file_fd);
+    ret=remove(FILE_PATH);
+    if(ret==-1){
+        perror("remove");
+        syslog(LOG_USER, "Error while removing file");
+    }
+    while(SLIST_EMPTY(&head)!=0){
+        ele=SLIST_FIRST(&head);
+        pthread_join(ele->thread_id,NULL);
+        SLIST_REMOVE_HEAD(&head,entries);
+        free(ele);
+    }
+    if(buf!=NULL){
+    free(buf);
+    }
+    interrupted=true;   
+    exit (EXIT_SUCCESS);  
+    } 
+    if(signo == SIGALRM){
+        if(total_connections==0){
+        fd=open(FILE_PATH, O_RDWR | O_CREAT | O_APPEND ,0644);
+        write_time_to_file(fd);
```

Comment:
@Sona-Subair I think you could simplify the logic and write from one place, and using SIGEV_THREAD would be simpler since you could handle locking requirements from your timestamp writer.  Also I'm not seeing the logic which increments total_connections so I think the branch below is dead?